packages/ui-causal-graph-editor/src/shared/use-causal-graph-editor.tsx

```diff
@@ -93,6 +106,23 @@ export default function useCausalGraphEditor(
         }, {} as GraphApi);
     }, [dispatch]);
 
+    const isTimeSeriesCausalGraph = useMemo(() => {
+        const nodeClass = state.graph.getNodeAttribute(state.graph.nodes()[0], 'extras')?.node_class;
+        return nodeClass === 'TimeSeriesNode';
+    }, [state.graph]);
+
+    // If the graph is a time series graph, tiers are not defined and is not a PlanarLayout, we update the tiers to show the time series in layers
+    if (
+        isTimeSeriesCausalGraph &&
+        'tiers' in graphLayout &&
+        'orientation' in graphLayout &&
+        graphLayout.tiers === undefined &&
+        !(graphLayout instanceof PlanarLayout)
+    ) {
+        updateNodesForTimeSeries(state.graph, api);
```

Comment:
The difference between this code and the canonical example in https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes is that here we're mutating state within a class and not invoking `setState` anywhere so this won't have the same behaviour of restarting a new render.

Though I'm not too worried about the graph itself, since we're not really tying it to React UI deliberately - the engine is isolated so it's a bit of a special case.

What I would be worried about is that:
1) you're mutating a prop here (in the form of `graphLayout`) - see https://react.dev/learn/keeping-components-pure ; it would be fine if you created a modified clone of the layout and return it from here
2) this will rerun on every render it seems like? again in the example from effect docs this pattern works because within the if clause they are calling `setState` which makes the next re-render *not* go inside the if again, preventing circular renders 

I feel like we're conflating two things here:
1) updating the graphLayout based on some conditions -> this could be a useMemo and return a new graph layout here
2) adjusting the graph state to contain tiers based on some conditions -> this feels like it belongs in the parser or should be run together with the parser before updating the state - so in 1) the reducer initializer, and 2) in the effect which invokes INIT_GRAPH before

I understand the condition might be the same for both (and we can extract it to a helper or something) but these are fundamentally different things - 1) is an adjustment to the graph state which should be done where the state is put into the reducer state, and 2) is an adjustment of graph layout which is not stored anywhere and just returned, so it can be constructed and returned (and wrapped in a useMemo but that's an optimization)

I hope this makes sense, let me know and we can chat about it more. 

Also cc @Roman-Kornev, feel free to chime in if you have some thoughts on how to not turn it into more of a spaghetti than it already is (also my fault) 😅

Reply:
Ok I see what you mean, will give it another go with the parser, that does seem like it might be a good place for it to be added

Reply:
Agreed, `graphLayout` is just derived state so it should go into a `useMemo`.

The trouble with `updateNodesForTimeSeries` also lies in the fact that it only updates in 1 direction: from non-timeseries to timeseries. Can we convert the graph from timeseries back to non-timeseries? In that case those extras would stay, but should they? All those checks should probably go in the parser, yeah.

Reply:
It is not updating a non time series to time series graph, it is just adding another property if it is a time series graph so that we can use internally to assign the layers.